www/assets/js/sofia.user.js

```diff
@@ -3,31 +3,82 @@ var S = S || {};
+
+          if(last.db !== S.db.remoteDB._db_name){
+            return false; //Not same remote DB
+          }
+
+          if(last.user.name !== S.config.user.username || last.user.pass !== S.config.user.userpass){
+            return false; //Not same cred
+          }
+
+          var ThreeHourSooner = new Date();
+          ThreeHourSooner.setHours(ThreeHourSooner.getHours() - 3);
+          if(last.at < ThreeHourSooner){
+            return false; //Not logged in since 3 Hours
+          }
+
+          return true; //All good
+
+        },
         isLogged : function(){
             return (typeof S.user._current.name !== "undefined" && (S.user._current.name !== null || S.user._current.roles[0] === "_admin" ) ) ;
-            //TODO rellogin if not based on registred value
-            /*
-            //TODO don't use jquery promise ?
-            var deferred = new $.Deferred()
-            db.getSession(function (err, response) {
-              if (err) {
-                // network error
-                deferred.reject("No network !");
-              } else if (!response.userCtx.name) {
-                // nobody's logged in
-                deferred.reject("Not logged !");
-              } else {
-                // response.userCtx.name is the current user
-                deferred.resolve(response.userCtx);
-              }
-            });
-            return deferred.promise();
-            */
+        },
+        isAdmin : function(){
+            console.log("isAdmin ? ", S.user._current.isLogged && ($.inArray("_admin", S.user._current.roles)!==-1 || $.inArray("sofia-admin", S.user._current.roles)!==-1));
+            return S.user._current.isLogged && ($.inArray("_admin", S.user._current.roles)!==-1 || $.inArray("sofia-admin", S.user._current.roles)!==-1);
         }
     },
-    login : function(user,pass){
-      //TODO
-      console.log("Trying to login : ", user,pass);
-      return S.db.users.login(user,pass);
+    login : function(user,pass,silent){
+      console.log("Trying to login : ", user,pass,silent);
+      if(typeof silent === "undefined"){
+        silent = false;
+      }
+      return S.db.users.login(user,pass,silent).then(function(doc){
+        //We are logged in
+        localStorage["sofia-last-login"] = JSON.stringify({
+            db :  S.db.remoteDB._db_name,
+            user : {
+              name : user,
+              pass : pass
+            },
+            doc : doc,
+            at :  new Date()
+          });
+        return doc;
+      });
+    },
+    set : function(user,pass,doc){
+      console.log("Setting user in memory", user, pass, doc);
+      S.config.user = {
+          username : user,
+          userpass : pass
+      };
+      localStorage["sofia-user-config"] = JSON.stringify(S.config.user);
+      if(typeof doc !== "undefined"){
+        $.extend(S.user._current, doc);
+      }
+    },
+    logout : function(){
+      S.user._current.name !== null; //logout
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/sapk/SOFIA/file/2103351212/issues/source?bid=2990560&fileBranchId=3143500#l76)